### PR TITLE
Feature :  Add image upload support for chat messages

### PR DESCRIPTION
**Pull request by MathieuPFR**
## Summary
  Adds backend support for uploading and sending images in chat conversations.

  ## Changes
  - **Repository Layer**: New `uploadChatImage()` method uploads images to Firebase Storage
  - **ViewModel Layer**: `uploadAndSendImage()` handles full upload flow with loading states
  - **Storage**: Images stored at `conversations/{conversationId}/images/{messageId}.jpg`
  - **Image Processing**: Reuses existing `ImageProcessor` for compression/optimization
  - **Message Type**: Uses existing `MessageType.IMAGE` enum with download URL in content field

 ## How It Works:

  User selects image → uploadAndSendImage() →
    Repository.uploadChatImage() →
      ImageProcessor.processImage() →
      Upload to Storage →
      Get download URL →
    Create Message(type=IMAGE, content=downloadURL) →
    Add to chat

  ## Testing
  - Test coverage for success, failure, and edge cases
  - All existing tests passing
  - Tests were made with help of ChatGPT AI

  ## Next Steps
  UI implementation needed:
  - Camera capture launcher
  - Gallery picker integration
  - Display IMAGE type messages in chat

### CODE REVIEW

File: app/src/main/java/com/android/joinme/model/chat/ChatRepository.kt [hunk 1]
```diff
@@ -61,4 +63,24 @@ interface ChatRepository {
...
+   * The image is processed (compressed, oriented, resized) before upload to optimize storage and
+   * bandwidth. The resulting download URL should be stored in the Message content field.
+   *
+   * @param context Android context for accessing the content resolver.
+   * @param conversationId The unique identifier of the conversation.
+   * @param messageId The unique identifier for the message that will contain this image.
+   * @param imageUri The URI of the image to upload (from camera or gallery).
+   * @return The Firebase Storage download URL for the uploaded image.
+   * @throws Exception if the upload fails.
+   */
```
Comment (BryanDMP) [inline app/src/main/java/com/android/joinme/model/chat/ChatRepository.kt]: **Minor:** The KDoc is well-written, but the `@throws Exception` is too generic. It would be more helpful to document specific exception types that callers should handle (e.g., `IOException`, `SecurityException`, storage quota exceeded, etc.).

**Suggestion:** Consider documenting more specific exception scenarios, or at minimum mention that the exception wraps the underlying storage/processing error.
```kotlin
/**
 * ...previous KDoc
 *
 * @throws Exception if the upload fails due to network issues, storage quota,
 *         invalid image format, or file access problems. The exception message
 *         contains details about the specific failure.
 */
```

File: app/src/main/java/com/android/joinme/model/chat/ChatRepositoryRealtimeDatabase.kt [hunk 2]
```diff
@@ -138,6 +148,41 @@ class ChatRepositoryRealtimeDatabase(private val database: FirebaseDatabase) : C
...
+              .child(conversationId)
+              .child("images")
+              .child("$messageId.jpg")
+
+      val uploadTask = storageRef.putBytes(processedImageBytes).await()
+      Log.d(TAG, "Image uploaded successfully, getting download URL")
+
+      // Get the download URL
+      val downloadUrl = storageRef.downloadUrl.await().toString()
+      Log.d(TAG, "Chat image uploaded successfully: $downloadUrl")
```
Comment (BryanDMP) [inline app/src/main/java/com/android/joinme/model/chat/ChatRepositoryRealtimeDatabase.kt]: **Issue:** `Log.d` (debug logging) should not be used in production code. Only `Log.e` and `Log.w` are accepted for production logging, as debug logs add noise, impact performance, and may leak sensitive information.

**Suggestion:** Remove the debug log statement, or if the information is critical for diagnosing production issues, elevate to `Log.w` with appropriate justification.

File: app/src/main/java/com/android/joinme/ui/chat/ChatViewModel.kt [hunk 3]
```diff
@@ -275,4 +283,77 @@ class ChatViewModel(
...
+        // Generate message ID first
+        val messageId = chatRepository.getNewMessageId()
+
+        // Upload image with timeout
+        val downloadUrl =
+            withTimeout(IMAGE_UPLOAD_TIMEOUT_MS) {
+              chatRepository.uploadChatImage(context, currentConversationId, messageId, imageUri)
+            }
+
+        Log.d(TAG, "Image uploaded successfully, creating message")
```
Comment (BryanDMP) [inline app/src/main/java/com/android/joinme/ui/chat/ChatViewModel.kt]: **Issue:** `Log.d` (debug logging) should not be used in production code. Only `Log.e` and `Log.w` are accepted for production logging, as debug logs add noise, impact performance, and may leak sensitive information.

**Suggestion:** Remove the debug log statement.
```kotlin
// Remove this line:
Log.d(TAG, "Image uploaded successfully, creating message")
```

File: app/src/main/java/com/android/joinme/ui/chat/ChatViewModel.kt [hunk 4]
```diff
@@ -275,4 +283,77 @@ class ChatViewModel(
...
+      } catch (e: kotlinx.coroutines.TimeoutCancellationException) {
+        val errorMsg = "Image upload timeout. Please check your connection and try again."
+        Log.e(TAG, errorMsg, e)
+        _uiState.value = _uiState.value.copy(imageUploadError = errorMsg)
+        onError(errorMsg)
+      } catch (e: Exception) {
+        val errorMsg = "Failed to upload image: ${e.message}"
+        Log.e(TAG, errorMsg, e)
+        _uiState.value = _uiState.value.copy(imageUploadError = errorMsg)
+        onError(errorMsg)
```
Comment (BryanDMP) [inline app/src/main/java/com/android/joinme/ui/chat/ChatViewModel.kt]: **Issue:** The timeout exception is caught and handled separately from other exceptions, which is correct. However, both branches set `imageUploadError` in the state AND call `onError()`. This could lead to duplicate error handling in the UI if it listens to both.

**Suggestion:** Document the expected UI behavior, or consider having `onError` be the sole notification mechanism (remove state update) or vice versa. Actually both things have the same goal, so there is a duplication here.

**Review by BryanDMP — CHANGES_REQUESTED** [review]
**Overall Review Summary**

Good work on this PR! The image upload feature is well-structured and follows existing patterns in the codebase. The code is clean, readable, and correctly reuses `ImageProcessor` for image processing rather than duplicating logic.

**Key changes required:**

1. **Remove all `Log.d` statements**: Only `Log.e` and `Log.w` are accepted in production code (see inline comments on ChatRepositoryRealtimeDatabase.kt and ChatViewModel.kt).

2. **Extract constants**:  The `"images"` path segment should be a constant like `IMAGES_PATH` for consistency with other path constants.

3. **Import cleanup**:  Import `TimeoutCancellationException` at the top of ChatViewModel.kt instead of using the fully qualified name inline.

4. **Testability improvement**:  Consider removing the default value for `FirebaseStorage` parameter and injecting `ImageProcessor` as a dependency.

**Minor considerations (documented in inline comments):**
- Dual error notification (state + callback): clarify expected UI behavior

**Test files:**
The tests provide solid coverage of the new functionality. I've suggested merging a few tests that share identical setup to optimize the coverage/CI-time ratio. I used AI tools to assist with both the test optimization analysis and generating code suggestions for the inline comments to make addressing these changes easier for you.

Overall, this is solid work, just a few cleanups needed before approval.
